✨  Implement `RulesFor` for `GlobalAuthorizer` and `LocalAuthorizer` to enable `SelfSubjectRulesReview`

embik opened the pull request:
<!--
Thanks for creating a pull request!

If this is your first time, please make sure to review CONTRIBUTING.MD.

Please copy the appropriate `:text:` or icon to the beginning of your PR title:

:sparkles: ✨ feature
:bug: 🐛 bug fix
:book: 📖 docs
:memo: 📝 proposal
:warning: ⚠️ breaking change
:seedling: 🌱 other/misc
:question: ❓ requires manual review/categorization

-->
## Summary

This PR implements the stub `RulesFor` methods for both `GlobalAuthorizer` and `LocalAuthorizer` to enable usage of the `SelfSubjectRulesReview` API. I'll be completely honest, this felt "too easy", so I hope I'm doing the correct thing here. Output of `kubectl auth can-i --list` looked sensible to me.

Basically, I've aligned the way both the global and the local authorizer generate their cluster-specific authorizer and just called `RulesFor` on the scoped authorizer created from it. After PR review, I've also added two e2e test cases that make sure that the `SelfSubjectRulesReview` API gives expected responses.

On a minor note, this PR cleans up the `OWNERS` file because it's from the before times (unless we want to keep access to this package limited, which might be fair).

## Related issue(s)

Fixes #1924

## Release Notes

<!--
Please add a release note in the block below. Leave NONE only if no user-facing changes are in this PR.
-->

```release-note
Implement `SelfSubjectRulesReview` API, enabling usage of e.g. `kubectl auth can-i --list`
```

sttts commented on pkg/authorization/local_authorizer.go:
```diff
@@ -33,26 +33,38 @@ import (
-	roleLister               rbacv1listers.RoleClusterLister
-	roleBindingLister        rbacv1listers.RoleBindingClusterLister
-	clusterRoleBindingLister rbacv1listers.ClusterRoleBindingClusterLister
-	clusterRoleLister        rbacv1listers.ClusterRoleClusterLister
```
these were intentional I believe to ensure the informers are started. If you move them into the hot path, the `Lister()` call might come to later after the shared informer factory has been started. Creates ugly races.

mjudeikis replied:
code comment would be good if this is the case

embik replied:
Yeah, let me bring this back and add a note. I wasn't aware of this and thought they were just written at different times and there was room to unify the code.

embik replied:
I'll do a little PR history search if I can find some additional context.

sttts replied:
we do similar things in the controllers. They all store listers in their struct to enforce creation of the informers at constructor time.

embik replied:
Question here: Should we adjust the `GlobalAuthorizer` as well then? Because I copied this pattern from it, so it would be susceptible to the same races, no? Is there something with the `localKubeInformers` and `globalKubeInformers` passed to _that_ that makes it okay?

embik replied:
Deleted my previous comment by mistake. I copied this pattern over from the `GlobalAuthorizer`, which also takes `SharedInformerFactory` input parameters. Should we adjust that one as well, or is there something special with it that will prevent those races mentioned?

sttts commented on pkg/authorization/global_authorizer.go:
```diff
@@ -58,8 +58,13 @@ func NewGlobalAuthorizer(localKubeInformers, globalKubeInformers kcpkubernetesin
+	}
+
+	scopedAuth := a.newAuthorizer(cluster.Name)
+	return scopedAuth.RulesFor(ctx, user, namespace)
```
How about a small e2e to ensure `SelfSubjectAccessReview` works?

embik replied:
Sounds sensible. Will take a look at that.

embik commented:
@sttts @mjudeikis ready for re-review.

embik commented:
/retest

looks like flakes ...? Let's see. Tests are not related I think, but maybe I broke something.

sttts commented on pkg/authorization/global_authorizer.go:
```diff
@@ -33,33 +34,42 @@ import (
-				&rbac.ClusterRoleBindingLister{Lister: globalKubeInformers.Rbac().V1().ClusterRoleBindings().Lister().Cluster(clusterName)},
-			)
-		},
+		globalRoleLister: globalKubeInformers.Rbac().V1().Roles().Lister(),
```
nit: globalRoles

Would be less noisy.

embik replied:
I tried to keep it consistent with `LocalAuthorizer`, which looks like this:

```go
a := &LocalAuthorizer{
    // listers are saved in the struct here to ensure that informers are started and we do not encounter race conditions with them.
    roleLister:               versionedInformers.Rbac().V1().Roles().Lister(),
    roleBindingLister:        versionedInformers.Rbac().V1().RoleBindings().Lister(),
    clusterRoleLister:        versionedInformers.Rbac().V1().ClusterRoles().Lister(),
    clusterRoleBindingLister: versionedInformers.Rbac().V1().ClusterRoleBindings().Lister(),
}
```

sttts commented on pkg/authorization/local_authorizer.go:
```diff
@@ -41,6 +42,7 @@ type LocalAuthorizer struct {
 
 func NewLocalAuthorizer(versionedInformers kcpkubernetesinformers.SharedInformerFactory) (authorizer.Authorizer, authorizer.RuleResolver) {
 	a := &LocalAuthorizer{
+		// listers are saved in the struct here to ensure that informers are started and we do not encounter race conditions with them.
```
nit: are instantiated early and ... with starting them

embik replied:
Much clearer, thanks. Updated the comments.

sttts commented:
Some nits.

/lgtm
/approve
/hold

if you want to address the nits.

embik commented:
/hold cancel

Addressed comment nit. I'm not 100% sure how to proceed with the lister fields, but from my perspective it's okay, so this can go in with an lgtm.

sttts commented:
/lgtm
